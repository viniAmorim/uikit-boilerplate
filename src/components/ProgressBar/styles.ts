import styled from 'styled-components';

import { ColorOptions } from '../types';

export type StyledProgressBarProps = {
  height?: number;
  backgroundColor?: ColorOptions;
  foregroundColor?: ColorOptions;
  value: number;
  maxValue: number;
};

export const ProgressBarContainer = styled.div<
  Omit<StyledProgressBarProps, 'value' | 'maxValue'>
>`
  width: 100%;
  height: ${(props) => props.height}px;
  border-radius: 8px;
  background-color: ${({ theme, backgroundColor = 'grey100' }) =>
    theme.colors[backgroundColor]};
`;

export const Progress = styled.div<StyledProgressBarProps>`
  height: 100%;
  width: ${(props) => (props.value / props.maxValue) * 100}%;
  background-color: ${({ theme, foregroundColor = 'green500' }) =>
    theme.colors[foregroundColor]};
  border-radius: ${(props) =>
    (props.value / props.maxValue) * 100 === 100 ? '8px' : '8px 0px 0px 8px'};
`;
