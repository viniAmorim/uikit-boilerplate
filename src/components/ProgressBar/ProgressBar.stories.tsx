import React from 'react';

import { Story, Meta } from '@storybook/react';

import ProgressBar from './ProgressBar';
import { StyledProgressBarProps } from './styles';

export default {
  title: 'Atoms/ProgressBar',
  component: ProgressBar,
} as Meta;

const Template: Story<StyledProgressBarProps> = (args) => (
  <ProgressBar {...args} />
);

export const Zero = Template.bind({});
Zero.args = {
  maxValue: 100,
  value: 0,
  height: 5,
};

export const Half = Template.bind({});
Half.args = {
  maxValue: 100,
  value: 50,
  height: 5,
};

export const Full = Template.bind({});
Full.args = {
  maxValue: 100,
  value: 100,
  height: 5,
};

export const Red = Template.bind({});
Red.args = {
  maxValue: 100,
  value: 75,
  height: 5,
  foregroundColor: 'red500',
};
