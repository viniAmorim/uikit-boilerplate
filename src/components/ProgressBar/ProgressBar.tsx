import React, { forwardRef } from 'react';

import * as Styled from './styles';

const ProgressBar = forwardRef<HTMLDivElement, Styled.StyledProgressBarProps>(
  function ProgressBar(
    { value, maxValue, height = 20, backgroundColor, foregroundColor, ...rest },
    ref,
  ) {
    return (
      <Styled.ProgressBarContainer
        data-testid="progress-bar-container"
        height={height}
        backgroundColor={backgroundColor}
        ref={ref}
        {...rest}
      >
        <Styled.Progress
          data-testid="progress-bar-progress"
          value={value}
          maxValue={maxValue}
          foregroundColor={foregroundColor}
        />
      </Styled.ProgressBarContainer>
    );
  },
);

export default ProgressBar;
