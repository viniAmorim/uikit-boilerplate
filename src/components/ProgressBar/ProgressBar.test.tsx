import React from 'react';

import snapshotDiff from 'snapshot-diff';

import { render } from '~/test/utils';

import ProgressBar from './ProgressBar';

describe('ProgressBar', () => {
  it('should render correctly', () => {
    const { getByTestId } = render(<ProgressBar value={50} maxValue={100} />);
    const container = getByTestId('progress-bar-container');
    const progress = getByTestId('progress-bar-progress');

    expect(container).toBeInTheDocument();
    expect(progress).toBeInTheDocument();
    expect(container).toMatchSnapshot();
    expect(progress).toMatchSnapshot();
  });

  it('should updates progress', () => {
    const { rerender, asFragment } = render(
      <ProgressBar value={0} maxValue={100} />,
    );

    const initial = asFragment();

    rerender(<ProgressBar value={50} maxValue={100} />);

    const half = asFragment();

    expect(snapshotDiff(initial, half)).toMatchSnapshot();
  });

  it('should render custom props', () => {
    const { rerender, asFragment } = render(
      <ProgressBar value={50} maxValue={100} />,
    );

    const normal = asFragment();

    rerender(
      <ProgressBar
        value={50}
        maxValue={100}
        foregroundColor="black600"
        backgroundColor="pink400"
      />,
    );

    const custom = asFragment();

    expect(snapshotDiff(normal, custom)).toMatchSnapshot();
  });
});
