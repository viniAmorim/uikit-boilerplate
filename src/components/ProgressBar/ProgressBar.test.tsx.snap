// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar should render correctly 1`] = `
<div
  class="sc-beqWaB gnRzvg"
  data-testid="progress-bar-container"
  height="20"
>
  <div
    class="sc-gueYoa cQJnNl"
    data-testid="progress-bar-progress"
    value="50"
  />
</div>
`;

exports[`ProgressBar should render correctly 2`] = `
<div
  class="sc-gueYoa cQJnNl"
  data-testid="progress-bar-progress"
  value="50"
/>
`;

exports[`ProgressBar should render custom props 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB gnRzvg\\"
+     class=\\"sc-beqWaB gRsga-d\\"
      data-testid=\\"progress-bar-container\\"
      height=\\"20\\"
    >
      <div
-       class=\\"sc-gueYoa cQJnNl\\"
+       class=\\"sc-gueYoa hIRZwZ\\"
        data-testid=\\"progress-bar-progress\\"
        value=\\"50\\"
      />
    </div>
  </DocumentFragment>"
`;

exports[`ProgressBar should updates progress 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -3,11 +3,11 @@
      class=\\"sc-beqWaB gnRzvg\\"
      data-testid=\\"progress-bar-container\\"
      height=\\"20\\"
    >
      <div
-       class=\\"sc-gueYoa kpXHzU\\"
+       class=\\"sc-gueYoa cQJnNl\\"
        data-testid=\\"progress-bar-progress\\"
-       value=\\"0\\"
+       value=\\"50\\"
      />
    </div>
  </DocumentFragment>"
`;
